### PR TITLE
Feature/visibility

### PR DESCRIPTION
**Pull request by DzmitryI**
[DIMS](https://dims-f1a5f.web.app/Auth "link for app")
##### information: #####
- admin@mail.ru 123456
- mentor@mail.ru 123456

### CODE REVIEW

File: src/components/app/app.scss [hunk 1]
```diff
@@ -116,7 +116,7 @@ table {
   position: relative;
   width: 90%;
   margin: 0 auto;
-  padding: 30px 0;
+  // padding: 2rem 0;
```
Comment (maximsan) [inline src/components/app/app.scss:119]: remove

File: src/components/auth/Auth.js [hunk 2]
```diff
@@ -59,6 +69,17 @@ class Auth extends PureComponent {
...
     authInput[controlName].value = value;
     authInput[controlName].touched = true;
+    let isFormValid = true;
+    Object.keys(authInput).forEach((name) => {
+      isFormValid = authInput[name].valid && isFormValid;
+    });
+    this.setState({ authInput, isFormValid });
+  };
+
+  onHandleFinishEditing = (controlName) => (event) => this.handleFinishEditing(event, controlName);
```
Comment (maximsan) [inline src/components/auth/Auth.js:79]: wrap in brackects

File: src/components/auth/Auth.js [hunk 3]
```diff
@@ -67,8 +88,20 @@ class Auth extends PureComponent {
     this.setState({ authInput, isFormValid });
   };
 
-  handleRadioClick = ({ target: { value } }) => {
-    this.setState({ base: value });
+  onHandleFocus = (controlName) => () => this.handleFocus(controlName);
```
Comment (maximsan) [inline src/components/auth/Auth.js:91]: wrap in brackects

File: src/components/auth/Auth.js [hunk 4]
```diff
@@ -67,8 +88,20 @@ class Auth extends PureComponent {
...
+
+  handleFocus = (controlName) => {
+    const authInput = { ...this.state.authInput };
+    authInput[controlName].valid = true;
+    authInput[controlName].touched = true;
+    this.setState({ authInput });
+  };
+
+  onHandlelSelect = (controlName) => (event) => this.handleSelect(event, controlName);
```
Comment (maximsan) [inline src/components/auth/Auth.js:100]: wrap in brackects

File: src/components/membersGrid/MembersGrid.js [hunk 5]
```diff
@@ -58,38 +61,37 @@ class MembersGrid extends Component {
...
+                <div>
+                  <ButtonLink
+                    className='btn-progress'
+                    onClick={this.onProgressClick}
+                    name='Progress'
+                    to='/MemberProgressGrid'
+                  />
+                  <ButtonLink className='btn-tasks' onClick={this.onShowClick} name='Tasks' to='/MemberTasksGrid' />
+                </div>
+                {ADMIN === email && (
```
Comment (maximsan) [inline src/components/membersGrid/MembersGrid.js:89]: create isAdmin variable

File: src/components/registration/Registration.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      authInput: { email, password },
+    } = this.state;
+    this.props.auth(email.value, password.value, false);
+  };
+
+  submitHandler = (event) => {
+    event.preventDefault();
+  };
+
+  onHandlelInput = (controlName) => (event) => this.handleInput(event, controlName);
```
Comment (maximsan) [inline src/components/registration/Registration.js:47]: wrap in brackets

File: src/components/registration/Registration.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    });
+    this.setState({ authInput, isFormValid });
+  };
+
+  onHandleFocus = (controlName) => () => this.handleFocus(controlName);
+
+  handleFocus = (controlName) => {
+    const authInput = { ...this.state.authInput };
+    authInput[controlName].valid = true;
+    authInput[controlName].touched = true;
```
Comment (maximsan) [inline src/components/registration/Registration.js:77]: follow DRY

File: src/components/registration/Registration.js [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    this.setState({ authInput });
+  };
+
+  renderInputs() {
+    const { authInput } = this.state;
+    return Object.keys(authInput).map((controlName, index) => {
+      const control = authInput[controlName];
```
Comment (maximsan) [inline src/components/registration/Registration.js:84]: spread control

File: src/components/registration/Registration.js [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    const { isFormValid } = this.state;
+    if (isRegistred) {
+      return <Redirect push to='/Auth' />;
+    }
+    return (
+      <>
+        {onNotification && <DisplayNotification notification={notification} />}
+        <div className='auth'>
+          <div className='auth-img'>
+            <img src={imgLogo} with='100px' height='50px' alt='logo' />
```
Comment (maximsan) [inline src/components/registration/Registration.js:115]: create Image component

File: src/store/reducers/auth.js [hunk 10]
```diff
@@ -1,28 +1,41 @@
...
-        base,
+        database,
+      };
+    case AUTH_REGISTER:
+      return {
+        ...state,
+        token: null,
+        email: null,
+        database: null,
+        isRegistred: true,
```
Comment (maximsan) [inline src/store/reducers/auth.js:30]: create helper method

File: src/page/helpersPage.js [hunk 11]
```diff
@@ -21,6 +21,15 @@ function clearOblectValue(objInput, objElem) {
...
   }
+  if (objElemClear.startDate) {
+    objElemClear.startDate = new Date();
+  }
+  if (objElemClear.deadlineDate) {
+    objElemClear.deadlineDate = new Date();
+  }
+  if (objElemClear.birthDate) {
+    objElemClear.birthDate = new Date();
+  }
```
Comment (maximsan) [inline src/page/helpersPage.js:32]: simplify the method, it seems ugly. don't forget add unit tests

File: src/components/helpersComponents.js [hunk 12]
```diff
@@ -49,6 +49,13 @@ const SetUp = ({ fetchServiceValue, roleValue, ThemeValue, component }) => {
   );
 };
 
+const countAge = (value) => {
+  const curDate = new Date();
+  const birthDate = new Date(value);
+  const age = curDate.getFullYear() - birthDate.getFullYear();
+  return curDate.setFullYear(curDate.getFullYear()) < birthDate.setFullYear(curDate.getFullYear()) ? age - 1 : age;
+};
```
Comment (maximsan) [inline src/components/helpersComponents.js:57]: do you have unit tests ?

File: src/components/app/app.scss (hunk 1 above)
Comment (DzmitryI) [reply src/components/app/app.scss:119]: fixed

File: src/components/membersGrid/MembersGrid.js (hunk 5 above)
Comment (DzmitryI) [reply src/components/membersGrid/MembersGrid.js:89]: fixed

File: src/components/registration/Registration.js (hunk 8 above)
Comment (DzmitryI) [reply src/components/registration/Registration.js:84]: fixed

File: src/components/registration/Registration.js (hunk 9 above)
Comment (DzmitryI) [reply src/components/registration/Registration.js:115]: fixed

File: src/components/registration/Registration.js (hunk 6 above)
Comment (DzmitryI) [reply src/components/registration/Registration.js:47]: fixed

File: src/components/auth/Auth.js (hunk 4 above)
Comment (DzmitryI) [reply src/components/auth/Auth.js:100]: fixed

File: src/components/auth/Auth.js (hunk 3 above)
Comment (DzmitryI) [reply src/components/auth/Auth.js:91]: fixed

File: src/components/auth/Auth.js (hunk 2 above)
Comment (DzmitryI) [reply src/components/auth/Auth.js:79]: fixed

File: src/components/registration/Registration.js (hunk 7 above)
Comment (DzmitryI) [reply src/components/registration/Registration.js:77]: fixed

File: src/store/reducers/auth.js (hunk 10 above)
Comment (DzmitryI) [reply src/store/reducers/auth.js:30]: fixed

File: src/page/helpersPage.js (hunk 11 above)
Comment (DzmitryI) [reply src/page/helpersPage.js:32]: fixed

**Pull request merged**